File: leetcode/python/211_add_and_search_words.py
class WordDictionary:
    def __init__(self):
        self.trie = {}

    def addWord(self, word: str) -> None:
        lookup = self.trie
        for c in word:
            if c not in lookup:
                lookup[c] = {}
            lookup = lookup[c]
        lookup['#'] = True
        print(self.trie)

    def search(self, word: str) -> bool:
        return self.searchImpl(0, word, self.trie)

    def searchImpl(self, index: int, word: str, lookup: dict) -> bool:
        if index == len(word):
            return '#' in lookup
        
        c = word[index]
        if c != '.':
            if c not in lookup:
                return False
            if self.searchImpl(index+1, word, lookup[c]):
                return True
        else:
            for key in lookup.keys():
                if key != '#' and self.searchImpl(index+1, word, lookup[key]):
                    return True
        return False


# obj = WordDictionary()
# obj.addWord('bad')
# obj.addWord('dad')
# obj.addWord('mad')
# print(obj.search('pad'))
# print(obj.search('bad'))
# print(obj.search('b..'))
# print(obj.search('.ad'))


w = WordDictionary()
w.addWord('a')
w.addWord('a')
# print(w.search('.'))
# print(w.search('a'))
# print(w.search('aa'))
# print(w.search('a'))
# print(w.search('.a'))
print(w.search('a.'))

File: leetcode/c++/211_add_and_search_words.cpp
#include <string>
#include <unordered_map>
#include <memory>

using namespace std;

struct TrieNode {
    unordered_map<char, unique_ptr<TrieNode>> children;
    bool isEnd {false};
};

class WordDictionary {
public:
    WordDictionary() {
        
    }
    
    void addWord(string word) {
        TrieNode* node = root.get();
        for (char c : word) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = make_unique<TrieNode>();
            }
            node = node->children[c].get();
        }
        node->isEnd = true;
    }
    
    bool search(string word) {
        return searchImpl(0, root.get(), word);
    }
private:
    unique_ptr<TrieNode> root {make_unique<TrieNode>()};

    bool searchImpl(int i, TrieNode* node, string& word) {
        if (i == word.size())
            return node->isEnd;
        if (word[i] == '.') {
            for (auto& [c, child] : node->children) {
                if (searchImpl(i + 1, child.get(), word))
                    return true;
            }
            return false;
        } else {
            if (node->children.find(word[i]) == node->children.end()) {
                return false;
            }
            return searchImpl(i + 1, node->children[word[i]].get(), word);
        }
    }

};
